monitoring/src/supply-monitor.ts

```diff
@@ -0,0 +1,101 @@
+import { SystemEventType } from "./system-event"
+
+import type { BigNumber } from "ethers"
+import type { TBTCToken } from "@keep-network/tbtc-v2.ts/dist/src/chain"
+import type { Monitor as SystemEventMonitor, SystemEvent } from "./system-event"
+
+// The block span the supply change is checked for. It is 12 hours expressed
+// in ETH blocks, assuming 12 seconds per block on average.
+const totalSupplyBlockSpan = (12 * 60 * 60) / 12
+
+const totalSupplyChangeThreshold = 10 // 10%
+
+const HighTotalSupplyChange = (
+  difference: BigNumber,
+  change: BigNumber,
+  threshold: number,
+  referenceBlock: number,
+  currentBlock: number
+): SystemEvent => ({
+  title: "High TBTC token total supply change",
+  type: SystemEventType.Critical,
+  data: {
+    change: `${difference.gte(0) ? "+" : "-"}${change.toString()}%`,
+    threshold: `${threshold}%`,
+    referenceBlock: referenceBlock.toString(),
+  },
+  block: currentBlock,
+})
+
+export interface SupplyMonitorPersistence {
+  lastHighTotalSupplyChangeBlock: () => Promise<number>
+  updateLastHighTotalSupplyChangeBlock: (block: number) => Promise<void>
+}
+
+export class SupplyMonitor implements SystemEventMonitor {
+  private tbtcToken: TBTCToken
+
+  private persistence: SupplyMonitorPersistence
+
+  constructor(tbtcToken: TBTCToken, persistence: SupplyMonitorPersistence) {
+    this.tbtcToken = tbtcToken
+    this.persistence = persistence
+  }
+
+  async check(fromBlock: number, toBlock: number): Promise<SystemEvent[]> {
+    // eslint-disable-next-line no-console
+    console.log("running supply monitor check")
+
+    // By default, the supply change is checked between the current and
+    // reference blocks.
+    const currentBlock = toBlock
+    let referenceBlock =
+      currentBlock - totalSupplyBlockSpan > 0
+        ? currentBlock - totalSupplyBlockSpan
+        : 0
```

Comment:
I am confused about the reference block. If the time span between `fromBlock` and `toBlock` is less than 12h on average, are we checking between `fromBlock` and... `0`?

Could be worth dropping a comment about this case, ideally explaining it on a timeline.

Reply:
> If the time span between fromBlock and toBlock is less than 12h on average, are we checking between fromBlock and... 0?

Negative. This is just a sanity check to not nuke the monitoring if the reference block is <0. If the time span is less than 12h, the reference block is replaced by the last block the alert was fired on.

Reply:
Left an additional explanation: 567dd2dd

Reply:
I think I see it now. For this particular monitor, `fromBlock` is ignored and the `referenceBlock` replaces it. 

Can we add a small comment on the top of `referenceBlock` to explain it?

Reply:
faea9554